File: minimum-height-trees/minimum-height-trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        if n <= 1:
            return [0]
        g = defaultdict(list)
        for v, e in edges:
            g[v].append(e)
            g[e].append(v)
        
        q = [i for i in range(n) if len(g[i]) == 1]
        print(q)
        while n > 2:
            n -= len(q)
            l = []
            for v in q:
                nxt = g[v].pop()
                g[nxt].remove(v)
                if len(g[nxt]) == 1:
                    l.append(nxt)
            q = l
        return q
            
            

File: minimum-height-trees/minimum-height-trees.java
class Solution {
    Map<Integer, List<Integer>> graph = new HashMap<>();
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        if (edges.length == 0) return new ArrayList<Integer>(){{add(0);}};
        for (int[] e: edges) {
            graph.putIfAbsent(e[0], new ArrayList<>());
            graph.putIfAbsent(e[1], new ArrayList<>());
            graph.get(e[0]).add(e[1]);
            graph.get(e[1]).add(e[0]);
        }

        ArrayList<Integer> q = new ArrayList<>();
        for (int i=0;i<n;i++) {
            if (graph.containsKey(i) && graph.get(i).size() == 1) q.add(i);
        }

        ArrayList<Integer> nq;
        while (n > 2) {
            n -= q.size();
            System.out.println(q.size());
            nq = new ArrayList<>();
            for (int v: q) {
                int u = graph.get(v).remove(0);
                graph.get(u).remove(new Integer(v));
                if (graph.get(u).size() == 1) nq.add(u);
            }
            q = nq;
            System.out.println(nq.size());
        }
        return q;
    }
}